Support calendars other than Calendar.ISO

kipcole9 opened the pull request:
This PR adds support for calendars other than `Calendar.ISO`.

### Enhancements

* Adds a behaviour, `CalendarInterval` with three callbacks:

   * `add/3` used to increment `:years`, `:quarters` and `:months`.  It could conceivably be used to expand to `:weeks` although this is much more complex to generalise. 

   * `quarter/2` returns a `Date.Range.t` for a given `year` and `quarter`

   * `quarter_of_year/3` returns the quarter for a given `year`, `month` and `day`

* Adds a precision of `:quarter` for intervals

* Updates `CalendarInterval.next_ndt/3` and `CalendarInterval.prev_ndt/3` to call the new callbacks when the interval is in a calendar other than `Calendar.ISO`.

* Adds the ability to parse intervals with a calendar modifier.  For example, `CalendarInterval.parse!("2019")` parses as expected in the `Calendar.ISO` calendar.  `CalendarInterval.parse!("2019 Cldr.Calendar.Gregorian")` parses in a calendar called `Cldr.Calendar.Gregorian` which is assumed to exist.

* Adds the calendar string to `to_string/1` when the calendar is not `Calendar.ISO`

* Adds the package [ex_cldr_calendars](https://hex.pm/packages/ex_cldr_calendars) to the `:dev` and `:test` environments only to allow testing of alternative calendars. This is not a hard dependency, it is only used as a source of test calendars.

### Next steps

* Propose including `add/3` to the standard lib `Calendar` behaviour for `:years`, `:months`, `:days`.  In general the `Calendar` module is the only module aware of `Date`, `Time` and `DateTime` structures. So the current version of `<calendar>.add/3` leaks implementation because it is passed a `NaiveDateTime`. This is the only function in calendar modules that does. Adding this call would also allow the removal of calendar calculation code completely from `CalendarInterval` which would seem a good idea.

* Update `ex_cldr_calendars` to observe the `CalendarInterval` behaviour.  It already implements the functions but does not annotate with the `@behaviour`.

kipcole9 committed:
Support calendars other than Calendar.ISO

kipcole9 committed:
Add support for a precision of :quarter

kipcole9 commented:
@wojtekmach let me know if this is something you will consider?

wojtekmach commented:
Sorry Kip, this has fallen through the cracks, will get back to you by end of the week.

> On 12 Sep 2019, at 13:32, Kip Cole <notifications@github.com> wrote:
> 
> @wojtekmach <https://github.com/wojtekmach> let me know if this is something you will consider?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/wojtekmach/calendar_interval/pull/6?email_source=notifications&email_token=AAASSJZ5FPDAY3D4JVKFTOLQJISDJA5CNFSM4ISLR3UKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6RR62A#issuecomment-530784104>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAASSJ525ZXWRCC27Y76L33QJISDJANCNFSM4ISLR3UA>.
>

wojtekmach commented on lib/calendar_interval.ex:
```diff
@@ -204,33 +228,89 @@ defmodule CalendarInterval do
+  end
+
+  defp do_parse!(<<year::4-bytes, "-", q::utf8, quarter::utf8>>, Calendar.ISO = calendar)
+       when q in [?q, ?Q] and quarter in ?1..?4 do
```
does ISO 8601 handle quarters, e.g. is `2019-Q3/Q4` a valid representation? If it is not part of the spec, I'd prefer to start with just non-ISO support and add quarters later (either by making them first-class like here, or by coming up with extension points like on #4) - would you be ok with that?

kipcole9 replied:
It is not an ISO8601 format as you say.  However it is certainly a *de jure* format and quarters are a normal interval in business. I'd like to ask for it to be considered (and weeks too - which are an ISO8601 format) but I can understand wanting to defer that for now.  I will revert this commit.  It has at least helped my understand how to add different intervals to the code :-)

wojtekmach commented on lib/calendar_interval.ex:
```diff
@@ -714,11 +842,16 @@ defmodule CalendarInterval do
       member?(interval, ndt)
     end
 
+    def member?(interval, %Date{calendar: calendar} = date) do
```
fyi, I'm having 2nd thoughts about current behaviour (on master) that:

```elixir
iex>  ~I"2019-01-01" in ~I"2019-01/12"
true
```

because:

```elixir
iex> Enum.any?(~I"2019-01/12", & &1 == ~I"2019-01-01")
false
```

and so we arguably break semantics. Curious what's your take on this? Of course it's outside of the scope of this PR, but I thought I'd mention it.

kipcole9 replied:
Thats interesting and I hadn't thought about it. In Allen's interval terms its asking the question of whether one interval `starts`, `finishes`, `during` or `equals` the other. Maybe that's the semantics that could be applied?  

In your first example its striating the first interval as a scalar, in the second more like a vector. I think it should always be considered a vector - and hence using Allen's Interval semantics.  But its only my first coffee so maybe not fully thought out.

wojtekmach replied:
Just to be clear, what I meant was that my understanding of Elixir enumerable semantics is that: if element X is _in_ enumerable XS then if we list all elements of XS, X will be one of them (as in `===/2`) (and pardon for my bad attempt at trying to come up with a property :))

So, taking the example from comment above, we'd have the following:

```elixir
iex> ~I"2019-01" in ~I"2019-01/12"
true

iex> ~I"2019-01-01" in ~I"2019-01/12"
false

iex> CalendarInterval.enclosing(~I"2019-01-01", :month) in ~I"2019-01/12"
true

iex> CalendarInterval.<some_predicate>?(~I"2019-01/12", ~I"2019-01-01")
true
```

I have hard time coming up with <some_predicate> function name, perhaps a generic `includes?` isn't too bad.

or maybe it'd be enough for people to do:

```elixir
iex> CalendarInterval.relation(~I"2019-12-01", ~I"2019-01/12") in [:starts, :during, :finishes]
true
```

(it currently does not work because we can only calculate relation of intervals of the same precision, but I think we can loosen it up a bit)

In any case, first of all, sorry for dragging you into this discussion with myself :D Second of all, does the concern make sense? And what do you think about the changes to the API?

LostKobrakai replied:
I guess the core problem here is that `~I"2019-01" in ~I"2019-01/12"` is not "element in enumerable", but rather "interval contained in interval". It becomes even more tricky when using an example, where the concept of "element in enumerable" breaks down: 
`~I"2019-01" in ~I"2019-01-01/20"`
Would that result in `true` or `false`? It get's easier when we declare "x in y" returns `true` if x is contained in y, for the ease of it's api, but it's certainly different semantics than "element in enumerable". The elixir core equivalent would actually be:
```
> 1..5 in 1..10
false
```

LostKobrakai replied:
I guess sticking to `x in y` results in true if `x` is a subinterval of `y` with the same granularity of `y` is the most sane approach, for the reasons mentioned initially:
```
 > Enum.into(~I"2019-01/12", [])
[~I"2019-01", ~I"2019-02", ~I"2019-03", …]
```
There's no other sane return value for the above if a CalendarInterval by itself should be an `Enumerable` (and imho it should be). We have no good other ways to denote "month" or "years" in elixir without falling back to strings or using e.g. the first date of the greater granularity value.

wojtekmach replied:
> I guess sticking to x in y results in true **if** x is a subinterval of y with the **same granularity** of y is the most sane approach

(emphasis mine) yes, that's a good way to put it, I'm leaning this way too. Thanks for feedback!

wojtekmach commented on lib/calendar_interval.ex:
```diff
@@ -204,33 +228,89 @@ defmodule CalendarInterval do
     end
   end
 
+  defp parse_including_calendar(string) do
```
I assume this is so that we can do: `~I"2019-01-01 Julian"`, you had a proposal like this for core, did it go anywhere?

kipcole9 replied:
Yes, it is so a calendar module can be specified.   The `inspect` part of of the proposal was accepted and is in the master branch for Elixir 1.10. I'll check if the sigils were accepted - I can't remember for the moment.

wojtekmach replied:
yeah, the inspect part was accepted but looks like the sigil not yet.

Regarding inspect, I like that it works by being able to implement `inspect_date` etc callbacks on the calendar implementation.

Should we follow that approach here as well - there would be a similar  `inspect_interval` callback? This might be useful for iso week calendar I guess?

wojtekmach commented on lib/calendar_interval.ex:
```diff
@@ -204,33 +228,89 @@ defmodule CalendarInterval do
+  defp parse_including_calendar(string) do
+    case String.split(string, " ", trim: true) do
+      [string] ->
+        [string, Calendar.ISO]
```
nitpick, if it always returns a 2 element list, let's use a 2-element tuple.

kipcole9 replied:
Done in [this commit](https://github.com/kipcole9/calendar_interval/commit/2b147c5d1e5952f2a334d9ad2fca9352de7c4ba6)

wojtekmach commented on lib/calendar_interval.ex:
```diff
@@ -32,6 +33,29 @@ defmodule CalendarInterval do
+  `NaiveDateTime.add/3`.
+  """
+  @callback add(Calendar.naive_datetime(), precision(), step :: integer()) ::
+              {Calendar.year(), Calendar.month(), Calendar.day()}
```
On principle, I'm not a fan of `add` (or +) operation on dates because we can run into this unfortunate situation:

```
~D[2017-03-31] + "2 months"  => ~D[2017-05-31]
                                           ^^

~D[2017-03-31] + "1 month"   => ~D[2017-04-30]
~D[2017-04-30] + "1 month"   => ~D[2017-05-30]
                                           ^^
```

so we break commutativity which we'd otherwise expect on a + operation. However, I guess 1. it's a callback and not public api, and, more importantly, 2. the semantics of the library are that you can only do e.g.: "add 1 X to date with X precision" (X=second, day, month, etc) so I guess we don't suffer from the mismatch shown above.

So I think the callback is fine if it helps other calendars, but just wanted to valid my assumptions. @kipcole9 am I making sense?

> Propose including add/3 to the standard lib Calendar behaviour for :years, :months, :days.

What would it be used by if added to core, are you thinking about public API like: `Date.shift(date, days: 2)`?

kipcole9 replied:
Yes adding to dates is tricky, however:

1. You're already doing it for the `Calendar.ISO` calendar - required of course to support `next/2` and `prev/2`. All this PR does is formalise the notion of adding a time period to a date as a `@callback`. 

2. And as you also note, the semantics of an interval with a precision actually avoids the issue you mention.  Its one of the beautiful things about intervals and why I look forward to getting rid of my own poor interval implementation and adopting yours.

Currently this PR has to have a different implementation because `Calendar.ISO` doesn't have the callbacks to supporting adding the necessary precision to an interval.  Each of the calendar structs has a form of `add/2` (`NaiveDateTime`, `Date` and `Time` but the semantics are different (days for `Date`, seconds for the others.  It would nice to harmonise this as a standard calendar callback.  But it doesn't have to be done - and therefore could be deferred as a discussion until later.  I certainly plan to declare the `CalendarInterval` `@behaviour` in my own calendars if you accept this PR.

wojtekmach replied:
> It would nice to harmonise this as a standard calendar callback.

right, my understanding is modules implementing `@behaviour CalendarInterval` may need this callback, so that's ok. Since you mentioned stdlib, I was just curious what would it be used for by modules implementing `@behaviour Calendar`.

> But it doesn't have to be done - and therefore could be deferred as a discussion until later.

I think keeping it as is is ok but deferring does sound good too. Your call!

wojtekmach reviewed (commented):
Great work @kipcole9. As I've written below, I'd see if we can move forward with smaller piece of the work which is just non-ISO handling and tackle quarters later.

wojtekmach commented:
Oh, and could you look into test failure?

wojtekmach commented on test/gregorian_calendar_test.exs:
```diff
@@ -0,0 +1,178 @@
+defmodule GregorianCalendarIntervalTest do
+  use ExUnit.Case
+  use CalendarInterval
+  alias Cldr.Calendar.Gregorian
```
to @LostKobrakai point (https://github.com/wojtekmach/calendar_interval/issues/5#issuecomment-523787141)

> I really like the idea of different calendars. I'd just like to see e.g. protocols or another optional way for sloting in other calendars. Setting up cldr is quite a lot to ask for if it's data is not really needed, so calendar_interval should probably work with and without it.

perhaps instead of bringing in ex_cldr_calendar for tests, which I think is totally fine btw, we should copy-paste [Calendar.Holocene](https://github.com/elixir-lang/elixir/blob/v1.9.1/lib/elixir/test/elixir/calendar/holocene.exs) into tests? The advantage would be that we'd see the implementation of the newly added callbacks, as opposed to relying on 3rd party package that does that under the hood.

If it's useful to keep ex_cldr_calendar for developing this integration, testing on different calendars etc, I'm fine with keeping the dep, but I'm beginning to think using Holocene for tests would make it easier to contribute to this project (because it's implementation is so trivial and included here). What do you think?

kipcole9 replied:
No problem at all to replace calendars.  I included this to demonstrate complete coverage of different kinds of calendars that are not simple shifts of `Calendar.ISO`.  I'm comfortable but the implementation in this PR covers the expectations of non`Calendar.ISO` calendars so a modified Holocene calendar for tests makes sense.  I'll update the PR. I will add the `CalendarInterval` `@behaviour` as well.

wojtekmach replied:
> I included this to demonstrate complete coverage of different kinds of calendars that are not simple shifts of Calendar.ISO.

Yeah, that's a good point. Maybe it's best to leave things as is then? Or have both something from ExCldr and the holocene (that's probably pushing it too much)? Not sure, your call!

kipcole9 replied:
I'll add an adapted Holocene calendar for testing as the next commit and remove the `ex_cldr_calendars` dependency - now that I think that @callback is in the right shape. I added a commit that changes the callback from `add/3` to `add/9` to be consistent with the practise that calendars don't know anything about the Date, Time, DateTime and NaiveDateTime structs.

kipcole9 commented:
> Oh, and could you look into test failure?

Curious - I don't have any test failures on my machine. The travis errors are about travis configuration and environment variables.

kipcole9 committed:
Revert "Add support for a precision of :quarter"

This reverts commit 3afb5ed18cd39ccede18be44f1901f7b3006f51d.

wojtekmach commented:
could you run `mix run examples/gaps.exs`?

kipcole9 commented:
> could you run `mix run examples/gaps.exs`?

Ahhhhhhh, now I see it :-)

kipcole9 committed:
Fix examples and refactor parse_including_calendar/1

kipcole9 committed:
Return a 2-tuple rather than list in parse_including_calendar/2

kipcole9 committed:
Change the add/3 callback to add/9
Calendars don't have any knowledge of the standard
structs of Date, Time, DateTime and NaiveDateTime.
So the callback uses the decomposed date/time
components.

kipcole9 commented:
At this point:

1. Quarters have been removed as a `:precision`
2. Return a 2-tuple not a list in `parse_including_calendar/1`
3. The `add` callback has been changed from `/3` to `/9` to be consistent with calendar practise
4. Fixed the bug causing the exception in the examples

To do:

1. Revisit the parsing with calendar since its not included in Elixir 1.10's sigils. Preserve the inspection approach of Elixir 1.10.

2. Add Holocene calendar for testing and remove `ex_cldr_calendars`

Am I missing any other "to do"'s?

wojtekmach commented:
Sounds good. I think we can keep sigil calendar parsing despite core not having it yet, hmm, so maybe ignore last comment about inspect_interval callback, because parsing requires certain shape. Sorry for back n forth, your call how to proceed with it.

kipcole9 committed:
Replace ex_cldr_calendars with built-in CalendarInterval.Holocene for tests

kipcole9 commented:
Removed `ex_cldr_calendars` from the test configuration and [added](https://github.com/kipcole9/calendar_interval/commit/68463d110c645da8ae9c22014cbc96d1f1f85d85) `CalendarInterval.Holocene` that implements the `CalendarInterval` `@behaviour`. 

I'd prefer to leave parsing with calendar names in since I will also continue to see if I can get that into Elixir core after 1.10 release :-)

I think this is now good to go?

kipcole9 commented:
Also formatted and ran dialyzer (without warnings so in the green).

wojtekmach committed:
Update CHANGELOG.md

wojtekmach committed:
Update README.md

wojtekmach commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## HEAD
```
@kipcole9 fyi, I'll leave this on master for a few days, I'm thinking about incorporating [1] change before next release.

If however, you'd like a release sooner than later to update your stuff, lmk!

[1] https://github.com/wojtekmach/calendar_interval/pull/6/files#r323873568

kipcole9 replied:
No rush @wojtekmach, I’ll work on deleting my not-very-good interval module in `ex_cldr_calendars` and replacing it with yours.  But there’s no immediate need.  I’ll think about the `interval in other_interval` issue as well now that my head is past this part.

wojtekmach committed:
Update mix.exs

wojtekmach committed:
Clean up holocene test

wojtekmach committed:
Update mix.exs

wojtekmach commented:
Thank you @kipcole9! <3

Fyi, I've landed some last-minute changes before merging, let me know what you think.

I've noticed minor problems with the holocene inspect/sigil implementation:

```elixir
$ MIX_ENV=test iex -S mix
iex> ~I"2019 CalendarInterval.Holocene"
~I"1201 CalendarInterval.Holocene"

iex> ~I"12019 CalendarInterval.Holocene"
** (FunctionClauseError) no function clause matching in CalendarInterval.do_parse!/2

    The following arguments were given to CalendarInterval.do_parse!/2:

        # 1
        "12019"
```

nothing we need to act on, it's just Holocene can have 5 digit years, where's ISO has 4. Something to think about when adding sigils to core though, that there might be a need for a "sigil callback" that understands non-iso format.

wojtekmach commented:
@kipcole9 regarding sigils, check out https://elixirforum.com/t/proposal-allow-calendar-to-be-specified-in-d-u-n-and-t-sigils/23156/18 :)
